core/common/src/main/java/alluxio/collections/IndexedSet.java

```diff
@@ -160,39 +155,42 @@ public void clear() {
    * {@code (o == null ? o2 == null : o.equals(o2))}. If this set already contains the object, the
    * call leaves the set unchanged.
    *
-   * @param objToAdd the object to add
-   * @return true if the object is successfully added to all indexes, otherwise false
+   * @param object the object to add
+   * @return true if this set did not already contain the specified element
    */
-  public boolean add(T objToAdd) {
-    Preconditions.checkNotNull(objToAdd);
-    synchronized (mLock) {
-      boolean success = mObjects.add(objToAdd);
-      // iterate over the first level index (filedid to map)
-      if (success) {
-        for (Map.Entry<FieldIndex<T>, Integer> index : mIndexMap.entrySet()) {
-          // get the second level map
-          Map<Object, Set<T>> fieldValueToSet = mSetIndexedByFieldValue.get(index.getValue());
-          // retrieve the key to index the object within the second level map
-          Object fieldValue = index.getKey().getFieldValue(objToAdd);
-          if (fieldValueToSet.containsKey(fieldValue)) {
-            success = fieldValueToSet.get(fieldValue).add(objToAdd);
-            if (!success) {
-              // this call can never return false because:
-              //   a. the second-level sets in the indices are all
-              //      {@link kava.util.HashSet} instances of unbounded space
-              //   b. We have already successfully added objToAdd on mObjects,
-              //      meaning that it cannot be already in any of the sets.
-              //      (mObjects is exactly the set-union of all the other second-level sets)
-              throw new IllegalStateException("Indexed Set is in an illegal state");
-            }
-          } else {
-            fieldValueToSet.put(fieldValue, Sets.newHashSet(Collections.singleton(objToAdd)));
-          }
-        }
+  public boolean add(T object) {
+    Preconditions.checkNotNull(object);
+
+    synchronized (object) {
```

Comment:
Perhaps document that this does not protect against concurrent `add` and `remove` of the same logical object but not actual object.